### ISSUE TITLE
Validation loss and accuracy is increasing while training accuracy is improving

### ISSUE THREAD

**Issue by morienor** [keras-team/keras#10102]
This is my first attempt at doing something with CNNs, so I am probably doing something very stupid - but can't figure out where I am wrong... 

The model seems to be learning fine, but the validation accuracy is not improving (ever - even after the first epoch), and validation loss is actually increasing with time. It doesn't look like I am overfiting (after 1 epoch?)  - must we off in some other way.

<img width="442" alt="screen shot 2018-05-03 at 01 07 20" src="https://user-images.githubusercontent.com/12259440/39555097-924d025e-4e6e-11e8-8579-3722ddba3e8e.png">

I am training a CNN network - I have ~100k images of various plants (1000 classes) and want to fine-tune ResNet50 to create a muticlass classifier. Images are of various sizes, I load them like so:

``` python
from keras.preprocessing import image                  

def path_to_tensor(img_path):
    # loads RGB image as PIL.Image.Image type
    img = image.load_img(img_path, target_size=(IMG_HEIGHT, IMG_HEIGHT))
    # convert PIL.Image.Image type to 3D tensor with shape (IMG_HEIGHT, IMG_HEIGHT, 3)
    x = image.img_to_array(img)
    # convert 3D tensor to 4D tensor with shape (1, IMG_HEIGHT, IMG_HEIGHT, 3) and return 4D tensor
    return np.expand_dims(x, axis=0)

def paths_to_tensor(img_paths):
    list_of_tensors = [path_to_tensor(img_path) for img_path in img_paths] #can use tqdm(img_paths) for data
    return np.vstack(list_of_tensors)
```

The database is large (does not fit into memory) and had to create my own generator to provide both reading from the disk and augmentation. (I know Keras has .flow_from_directory() - but my data is not structured this way - it is just a dump of 100k images mixed with 100k metadata files).

The generator version below doesn't do any augmentation for the time being - just rescaling:

``` python

def generate_batches_from_train_folder(images_to_read, labels, batchsize = BATCH_SIZE):    

#Generator that returns batches of images ('xs') and labels ('ys') from the train folder
#:param string filepath: Full filepath of files to read - this needs to be a list of image files
#:param np.array: list of all labels for the images_to_read - those need to be one-hot-encoded
#:param int batchsize: Size of the batches that should be generated.
#:return: (ndarray, ndarray) (xs, ys): Yields a tuple which contains a full batch of images and labels. 

dimensions = (BATCH_SIZE, IMG_HEIGHT, IMG_HEIGHT, 3)

    train_datagen = ImageDataGenerator(
        rescale=1./255,
        #rotation_range=20,
        #zoom_range=0.2, 
        #fill_mode='nearest',
        #horizontal_flip=True
    )

    # needs to be on a infinite loop for the generator to work
    while 1:
        filesize = len(images_to_read)

        # count how many entries we have read
        n_entries = 0
        # as long as we haven't read all entries from the file: keep reading
        while n_entries < (filesize - batchsize):
            
            # start the next batch at index 0
            # create numpy arrays of input data (features) 
            # - this is already shaped as a tensor (output of the support function paths_to_tensor)
            xs = paths_to_tensor(images_to_read[n_entries : n_entries + batchsize])

            # and label info. Contains 1000 labels in my case for each possible plant species
            ys = labels[n_entries : n_entries + batchsize]

            # we have read one more batch from this file
            n_entries += batchsize
              
            #perform online augmentation on the xs and ys
            augmented_generator = train_datagen.flow(xs, ys, batch_size = batchsize)
            
            yield  next(augmented_generator)

```

This is how I define my model:

``` python

def get_model():
    
    # define the model
    base_model = ResNet50(input_shape=DIMENSIONS, weights='imagenet', include_top=False)

    # Freeze the layers which you don't want to train. Here I am freezing all but the last 2 layers 
    for layer in base_model.layers[:-2]:
        layer.trainable = False
    
    model_gen = models.Sequential()

    # Add the vgg convolutional base model
    model_gen.add(base_model)

    #Adding custom Layers 
    model_gen.add(layers.GlobalAveragePooling2D(name='avg_pool_2'))
    model_gen.add(layers.Dense(1000, activation='softmax'))
    
    # compile the model 
    model_gen.compile(
        loss='categorical_crossentropy',
        optimizer=optimizers.Adam(1e-3),
        metrics=['acc'])
    
    return model_gen

```

I then use the following function to do a 5 fold CV for the model:
The code below only looks at the first fold - because the other ones don't work either...

``` python

from sklearn.model_selection import train_test_split, StratifiedKFold
folds = list(StratifiedKFold(n_splits=5, shuffle=True, random_state=1).split(train_images, train_to_integer))
# train_to_integer is a dense representation of Y_train below. Y_train is a one-hot-encoded version

#train the model on each fold
history = []

for j, (train_idx, val_idx) in enumerate(folds):
    
    if j >=1: # check the first fold only here
        break
        
    print('\nFold ',j)
    X_train_cv_path = train_images[train_idx]
    y_train_cv = Y_train[train_idx,:]
   
    X_valid_cv = train_images[val_idx]
    y_valid_cv= Y_train[val_idx,:]
    
    load_name_weights = "./keras_models/model_fold_" + str(j) + "_weights.h5"
    save_name_weights = "./keras_models/model_fold_" + str(j) + "_weights.h5"
    
    #saves the model after each epoch if the validation loss decreased
    checkpoint = ModelCheckpoint(save_name_weights, monitor='val_acc', verbose=1,
                                 #save_best_only=True,
                                 save_weights_only=False,
                                 mode='auto', period=1)
    
    early = EarlyStopping(monitor='val_acc', min_delta = 0, patience=5, verbose=1, mode='auto')

    callbacks = [checkpoint,early]
    
    train_generator = generate_batches_from_train_folder(
        images_to_read = X_train_cv_path, 
        labels = y_train_cv, 
        batchsize = BATCH_SIZE)
    
    #print('generating validation data...')
    
    valid_generator = generate_batches_from_train_folder(
        images_to_read = X_valid_cv, 
        labels = y_valid_cv, 
        batchsize = BATCH_SIZE)
    
    #x_valid_cv = paths_to_tensor(X_valid_cv) # this takes paths to images and return the np array
    
    #valid_datagen = ImageDataGenerator(rescale=1./255)
    #valid_generator = valid_datagen.flow(x_valid_cv, y_valid_cv, batch_size=BATCH_SIZE)
        
        
    try:
        model = load_model(load_name_weights)
        print('model loaded : ',load_name_weights)
    
    except:
        
        print('model not loaded - generating from scratch')
        model = get_model()
        
    
    history_fold = model.fit_generator(
                train_generator,
                steps_per_epoch = len(X_train_cv_path) // BATCH_SIZE,
                epochs = 10,
                verbose = 1,
                validation_data = valid_generator,
                validation_steps= len(X_valid_cv) // BATCH_SIZE,
                callbacks = callbacks)
    
    history.append(history_fold.history)
    
    history_file = './temp_data/fold_at3_history_data'
    np.save(history_file,history,allow_pickle=True)
    
    print('finished the test')
```

I have done various things - like checked that my generator is actually not yielding the same batch over and over - checked that both train and valid generators are basically the same, but nothing seems to show where the problem is...

**Comment by drcrook1** [comment]
So I have a scenario where my loss is converging; but my validation loss is diverging as well.  This depends on what I'm training and how far into the epochs I am.  So for some scenarios it tends to start around epoch 150 but in other items it starts in epoch 50 or epoch 200.

I'm going to piggy back here a bit and just ask if there is a known callback we can use with our model.fit() function which can check if our validation loss is increasing and if so for how long and by how much.

Even more ideal would be to identify the epoch where it started to diverge and roll the weights back to that epoch.  

The issue is clearly an over-fitting one; In my scenario a callback would be great because I overfit through over-training.

Here is a print out of the losses during a second "walk".  There is significant transfer learning in my solution and this is a step where it continues to train but the data is very similar to the step prior to it.

Epoch 4/200
 - 2s - loss: 55.8033 - val_loss: 26.4454
Epoch 5/200
 - 2s - loss: 53.3749 - val_loss: 26.3017
Epoch 6/200
 - 3s - loss: 51.0523 - val_loss: 26.2110
Epoch 7/200
 - 2s - loss: 48.8308 - val_loss: 26.1701
Epoch 8/200
 - 2s - loss: 46.7058 - val_loss: 26.1759
Epoch 9/200
 - 2s - loss: 44.6733 - val_loss: 26.2252
Epoch 10/200
 - 2s - loss: 42.7294 - val_loss: 26.3153
Epoch 11/200
 - 2s - loss: 40.8699 - val_loss: 26.4434
Epoch 12/200
 - 2s - loss: 39.0916 - val_loss: 26.6070
Epoch 13/200
 - 2s - loss: 37.3903 - val_loss: 26.8036

**Comment by ersamo** [comment]
did you solve it  ?
